File: src/Core/MarkdownTreaty.fs
namespace Zeta.Core

open System.Collections.Immutable

/// **Markdown + Frontmatter Treaty (Priority 1)**
///
/// Implements a 4-language read/write treaty for parsing .md files.
/// Delimited by "---" at the start of the file, containing frontmatter
/// serialized as canonical YAML, mapping to a DynamicValue.Object.
/// If there is no frontmatter, it parses metadata as an empty Object [].
module MarkdownTreaty =

    /// Parse a Markdown string into metadata (DynamicValue.Object) and the remaining body string.
    /// Asserts strict canonical check on the frontmatter YAML.
    let parse (text: string) : Result<DynamicValue * string, string> =
        let text = if System.Object.ReferenceEquals(text, null) then "" else text
        if text.StartsWith("---") && (text.Length = 3 || text.[3] = '\n' || (text.[3] = '\r' && text.Length > 4 && text.[4] = '\n')) then
            let headerLen = if text.[3] = '\r' then 5 else 4
            
            let rec findClosing (index: int) =
                if index >= text.Length then
                    None
                else
                    let isNewline, newlineLen, nextIdx =
                        if text.[index] = '\n' then true, 1, index + 1
                        elif text.[index] = '\r' && index + 1 < text.Length && text.[index + 1] = '\n' then true, 2, index + 2
                        else false, 0, index + 1
                    if isNewline then
                        if nextIdx + 3 <= text.Length && text.Substring(nextIdx, 3) = "---" then
                            let tailIdx = nextIdx + 3
                            if tailIdx = text.Length then
                                Some (index, newlineLen, tailIdx)
                            elif text.[tailIdx] = '\n' then
                                Some (index, newlineLen, tailIdx + 1)
                            elif text.[tailIdx] = '\r' && tailIdx + 1 < text.Length && text.[tailIdx + 1] = '\n' then
                                Some (index, newlineLen, tailIdx + 2)
                            else
                                findClosing nextIdx
                        else
                            findClosing nextIdx
                    else
                        findClosing nextIdx

            match findClosing headerLen with
            | None ->
                Error "Unclosed frontmatter delimiter"
            | Some (closeStart, newlineLen, closeEnd) ->
                let yamlPart = text.Substring(headerLen, closeStart + newlineLen - headerLen)
                let bodyPart = text.Substring(closeEnd)
                match DynamicValue.fromYaml yamlPart with
                | Error err -> Error (sprintf "Failed to parse frontmatter YAML: %A" err)
                | Ok (DynamicValue.Object pairs as metadata) ->
                    Ok (metadata, bodyPart)
                | Ok _ ->
                    Error "Frontmatter must be a YAML map (Object)"
        else
            Ok (DynamicValue.Object [], text)

    /// Serialize metadata (DynamicValue.Object) and a body string into a Markdown string with frontmatter.
    /// If metadata is empty (Object []), frontmatter section is omitted entirely.
    let serialize (metadata: DynamicValue) (body: string) : Result<string, EncodeError> =
        match metadata with
        | DynamicValue.Object [] ->
            Ok body
        | DynamicValue.Object _ ->
            match DynamicValue.toYaml metadata with
            | Error err -> Error err
            | Ok yaml ->
                Ok (sprintf "---\n%s---\n%s" yaml body)
        | _ ->
            Error EncodeError.NonRepresentable

File: src/Core.CSharp.DynamicValue/MarkdownTreaty.cs
namespace Zeta.Core.CSharp;

using System;
using System.Collections.Generic;
using System.Collections.Immutable;

/// <summary>
/// Markdown + Frontmatter Treaty (Priority 1) for C#.
/// Parses and serializes markdown files with frontmatter.
/// </summary>
public static class MarkdownTreaty
{
    public static Result<(DynamicValue Metadata, string Body), string> Parse(string text)
    {
        text ??= "";
        if (text.StartsWith("---", StringComparison.Ordinal) && (text.Length == 3 || text[3] == '\n' || (text[3] == '\r' && text.Length > 4 && text[4] == '\n')))
        {
            int headerLen = text[3] == '\r' ? 5 : 4;

            if (!TryFindClosingDelimiter(text, headerLen, out int closeStart, out int newlineLen, out int closeEnd))
            {
                return new Result<(DynamicValue, string), string>.Err("Unclosed frontmatter delimiter");
            }

            string yamlPart = text.Substring(headerLen, closeStart + newlineLen - headerLen);
            string bodyPart = text.Substring(closeEnd);

            var fromYamlRes = DynamicValues.FromYaml(yamlPart);
            if (fromYamlRes is not Result<DynamicValue, DecodeError>.Ok okYaml)
            {
                var errRes = (Result<DynamicValue, DecodeError>.Err)fromYamlRes;
                return new Result<(DynamicValue, string), string>.Err($"Failed to parse frontmatter YAML: {errRes.Error}");
            }

            if (okYaml.Value is not DynamicValue.Object metadata)
            {
                return new Result<(DynamicValue, string), string>.Err("Frontmatter must be a YAML map (Object)");
            }

            return new Result<(DynamicValue, string), string>.Ok((metadata, bodyPart));
        }
        else
        {
            var emptyObj = new DynamicValue.Object(ImmutableArray<KeyValuePair<string, DynamicValue>>.Empty);
            return new Result<(DynamicValue, string), string>.Ok((emptyObj, text));
        }
    }

    private static bool TryFindClosingDelimiter(
        string text, int startIdx, out int closeStart, out int newlineLen, out int closeEnd)
    {
        closeStart = -1; newlineLen = -1; closeEnd = -1;
        int index = startIdx;
        while (index < text.Length)
        {
            bool isNewline = false; int curNewlineLen = 0; int nextIdx = index;
            if (text[index] == '\n')
            {
                isNewline = true; curNewlineLen = 1; nextIdx = index + 1;
            }
            else if (text[index] == '\r' && index + 1 < text.Length && text[index + 1] == '\n')
            {
                isNewline = true; curNewlineLen = 2; nextIdx = index + 2;
            }

            if (isNewline)
            {
                if (nextIdx + 3 <= text.Length && string.Equals(text.Substring(nextIdx, 3), "---", StringComparison.Ordinal))
                {
                    int tailIdx = nextIdx + 3;
                    if (tailIdx == text.Length || text[tailIdx] == '\n' || (text[tailIdx] == '\r' && tailIdx + 1 < text.Length && text[tailIdx + 1] == '\n'))
                    {
                        closeStart = index;
                        newlineLen = curNewlineLen;
                        closeEnd = tailIdx == text.Length ? tailIdx : (text[tailIdx] == '\n' ? tailIdx + 1 : tailIdx + 2);
                        return true;
                    }
                }
                index = nextIdx;
            }
            else index++;
        }
        return false;
    }

    public static Result<string, EncodeError> Serialize(DynamicValue metadata, string body)
    {
        if (metadata is not DynamicValue.Object obj)
        {
            return new Result<string, EncodeError>.Err(EncodeError.NotXmlRepresentable);
        }

        if (obj.Pairs.Length == 0)
        {
            return new Result<string, EncodeError>.Ok(body ?? "");
        }

        var toYamlRes = DynamicValues.ToYaml(obj);
        if (toYamlRes is not Result<string, EncodeError>.Ok okYaml)
        {
            var errRes = (Result<string, EncodeError>.Err)toYamlRes;
            return new Result<string, EncodeError>.Err(errRes.Error);
        }

        string result = $"---\n{okYaml.Value}---\n{body ?? ""}";
        return new Result<string, EncodeError>.Ok(result);
    }
}
